### ISSUE TITLE
`unused_parens` false positive with impl block

### ISSUE THREAD

**Issue by kadiwa4** [rust-lang/rust#107235]
### Code

```Rust
struct Something;
impl (<Something>::Item) {}
```

### Current output

```Shell
warning: unnecessary parentheses around type
 --> src/lib.rs:2:6
  |
2 | impl (<Something>::Item) {}
  |      ^                 ^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
2 - impl (<Something>::Item) {}
2 + impl <Something>::Item {}
  |

error[E0391]: cycle detected when computing type of `<impl at src/lib.rs:2:1: 2:24>`
 --> src/lib.rs:2:1
  |
2 | impl (<Something>::Item) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^
  |
note: ...which requires collecting inherent impls for `Something`...
 --> src/lib.rs:1:1
  |
1 | struct Something;
  | ^^^^^^^^^^^^^^^^
  = note: ...which requires finding all inherent impls defined in crate...
  = note: ...which again requires computing type of `<impl at src/lib.rs:2:1: 2:24>`, completing the cycle
note: cycle used when collecting item types in top-level module
 --> src/lib.rs:1:1
  |
1 | / struct Something;
2 | | impl (<Something>::Item) {}
  | |___________________________^

error[E0223]: ambiguous associated type
 --> src/lib.rs:2:7
  |
2 | impl (<Something>::Item) {}
  |       ^^^^^^^^^^^^^^^^^ help: use the fully-qualified path: `<Something as IntoIterator>::Item`

Some errors have detailed explanations: E0223, E0391.
For more information about an error, try `rustc --explain E0223`.
warning: `playground` (lib) generated 1 warning
error: could not compile `playground` due to 2 previous errors; 1 warning emitted
```

### Desired output

```Shell
Same thing without the unused_parens warning.
```

### Rationale and extra context

Without the parentheses, it would be equivalent to
```rust
impl<Something> ::Item {}
```
Note that `Something` is now a generic type and `::Item` is a standalone item path.

### Anything else?

rustc nightly 2023-01-22

off-topic: it's also interesting that the error E0391 (which I find rather unnecessary in this case) isn't emitted on stable 1.66.1.

**Comment by Arjentix** [comment]
I faced the same issue.
Also it's not only the lint problem, it's `rustc` interpretation problem.

## Example

Looks like in this example`rustc` interprets `(Something)` as just `Something` (while I expected it to be interpreted like a *tuple*) and raises the lint warning:

```rust
struct Something;

trait MyTrait {}

impl MyTrait for (Something) {} // warning: unnecessary parentheses around type
```

But it can be fixed by adding `,` which I find a very awkward solution:
```rust
struct Something;

trait MyTrait {}

impl MyTrait for (Something,) {}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=41b1bdb44d5f30d8baa97598a65d498f) to try.

**Comment by fmease** [comment]
Similarly here:

```rs
pub fn f<T: Iterator>() where (<T>::Item): Copy {} //~ WARN unnecessary parentheses around type
```

Without parentheses, you'd get:

```
error: generic parameters on `where` clauses are reserved for future use
  --> src/lib.rs:10:31
   |
10 | pub fn f<T: Iterator>() where <T>::Item: Copy {}
   |                               ^^^ currently unsupported
```

Or here:

```rs
#![feature(more_qualified_paths)]

fn scope<T: Iterator<Item = String>>() {
    for (<T>::Item { .. }) in [] {} //~ WARN unnecessary parentheses around pattern
}
```

Without parentheses, you'd get:

```
error: expected one of `async`, `gen`, `move`, `static`, `use`, `|`, or `||`, found `::`
 --> src/lib.rs:4:12
  |
4 |     for <T>::Item { .. } in [] {}
  |            ^^ expected one of 7 possible tokens

error[E0658]: `for<...>` binders for closures are experimental
 --> src/lib.rs:4:5
  |
4 |     for <T>::Item { .. } in [] {}
  |     ^^^^^^^
```
